iXBRLViewerPlugin/xhtmlserialize.py

```diff
@@ -13,29 +13,172 @@
+    ENTITIES = {
+        '>': '&gt;',
+        '<': '&lt;',
+        '"': '&quot;',
+        '&': '&amp;'
+        }
+
+    def __init__(self, fout, xml_declaration = True, assume_xhtml = True):
+        self.fout = fout
+        self.encoding = "utf-8"
+        self.xml_declaration = xml_declaration
+        self.assume_xhtml = assume_xhtml
+
+    def write(self, s):
+        self.fout.write(s.encode(self.encoding))
+
+    def prefix_sort(self, p):
+        return p if p is not None else '0'
+
+    def qname_for_node(self, node):
+        qname = etree.QName(node.tag)
+        if qname.namespace is None:
+            prefix = None
+        elif node.nsmap.get(node.prefix, None) == qname.namespace:
+            prefix = node.prefix
+        else:
+            prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace), key = self.prefix_sort)))
+        if prefix is None:
+            return qname.localname
+        return "%s:%s" % (prefix, qname.localname)
+
+    def qname_for_attr(self, tag, nsmap):
+        qname = etree.QName(tag)
+        if qname.namespace is None:
+            return qname.localname
+        if qname.namespace == XML_NS:
+            prefix = 'xml'
+        else:
+            prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace and p is not None), key = self.prefix_sort)))
+        return "%s:%s" % (prefix, qname.localname)
+
+    def is_selfclosable(self, n):
+        qname = etree.QName(n)
+        return (
+                qname.localname in XHTMLSerializer.SELF_CLOSABLE
+                and (
+                    qname.namespace == XHTML_NS
+                    or (self.assume_xhtml and qname.namespace is None)))
+
+    def escape_attr(self, s):
+        return re.sub(r'([<>"&\u0001-\u001F\u007F-\u009F])', lambda m: self.escape_char(m.group(0)), s)
+
+    def xmlns_declaration(self, prefix, uri):
+        if prefix is None:
+            return 'xmlns="%s"' % self.escape_attr(uri)
+        return 'xmlns:%s="%s"' % (prefix, self.escape_attr(uri))
+
+    def namespace_declarations(self, new_nsmap, cur_nsmap):
+        changed = set(p[0] for p in (set(new_nsmap.items()) ^ set(cur_nsmap.items())))
+        return sorted(self.xmlns_declaration(p, new_nsmap[p]) for p in changed)
+
+    def escape_char(self, c):
+        return self.ENTITIES.get(c, '&#x%2X;' % ord(c))
+
+    def write_escape_text(self, s, escape_mode):
+        if s is None:
+            return
+        if escape_mode == EscapeMode.STYLE:
+            # HTML does not understand XML escapes inside a style element.
+            # Don't escape ">".  Escaping isn't required by XML, but is common
+            # practice, so we do it elsewhere.
+            # "&" and "<" may only appear in a string or comment in CSS, so escape
+            # as if they're in a string.
+            self.write(re.sub(r'([<&])', lambda m: "\\%06X" % ord(m.group(1)), s))
```

Comment:
> `<` and `&` are escaped as if they are appearing within a CSS string using `\NNNNNN` hex escapes (this is the only place other than comments in which they can legally appear in CSS).

I am somewhat nervous this will not remain true (`&` in [CSS Nesting Module working draft](https://www.w3.org/TR/css-nesting-1/)); files using that are obviously not in the XHTML/HTML subset, but it seems wrong to break the XHTML styling.  Well, it feels a little off to write `\` escapes in a non-string context.

Reply:
Ah, I did wonder if CSS syntax might start using them at some point.  If it does, I think we're stuck for maintaining XHTML/HTML compatibility.

For now, I think the options are:

* CSS parse the contents to find out if we're in a string (and either raise an error, or warn if not)
* Don't escape them at all (and raise an error/warn if they occur)

Reply:
If a valid iXBRL input document contains `<style>div::before { content: '&amp;&lt;'; }</style>`, then the producer _already_ did not care about HTML compatibility or else they would have used `'\26\3c'`.  ixbrl-viewer currently does not make this worse (the output document is as compatible with HTML as the input document), and I don't think we need any code to improve that, so we can just use `escape_char`.

However, if a valid iXBRL input document uses `<style>div > span { color: red; }</style>`, then the iXBRL producer might have been trying to care about HTML compatibility.  ixbrl-viewer currently breaks that by outputting `&gt;`, and I am interested in fixing this case because we have third party iXBRL producers that output styles like that.  I think a regex like `r'([<&]|]]>)'` could be used to fix those cases, including the extreme edge case of `<style>div::before { content: ']]&gt;' }</style>`.

I think the PR is currently missing handling of control character escapes in `<style>` for XML 1.0.  I don't know if there's a use case for a literal control character in a string in `<style>`, but I guess we could conditionally escape the control characters based on whether the XML version is 1.0.

To summarize, I suggest:
- Remove the special handling of `style` altogether.
- Add `]]>` to the regex.  (I guess `escape_char`/`ENTITIES` will be weirdly named, so perhaps remove the function altogether and just use an `ESCAPES` dict directly.)
- In addition to the escape padding fix and \t\r\n improvement discussion, consider only conditionally escaping control characters based on 1.0.

Reply:
> * Remove the special handling of `style` altogether.
> * Add `]]>` to the regex. 

To be clear, are you proposing that we consistently _don't_ escape `>` for all elements, unless it appears as part of `]]>` ?   I'm fine with that, but it's more of a change from the current behaviour.  I don't know of any good reason to escape `>`, but it does seem to be common practice.

> * In addition to the escape padding fix and \t\r\n improvement discussion, consider only conditionally escaping control characters based on 1.0.

Just to confirm, we're talking about the 7F-9F control characters here?

Reply:
> To be clear, are you proposing that we consistently don't escape `>` for all elements, unless it appears as part of `]]> `?

That was my suggestion since it would remove all need to track `style`, but if you prefer to keep tracking `style` just for that divergence, that's fine with me.

> Just to confirm, we're talking about the 7F-9F control characters here?

Sorry, ignore that final bullet, I misremembered: XML 1.1 added support for 0x01-0x1f but only as character references, not literals.

Reply:
Done, but I've retained the escape mode tracking.  In the past we had a request to create a single-file viewer (JS embedded), which could also use this escape-mode tracking for `script` tags, if implemented, and this behaviour also minimises change from what we had before (although I can't see how it would make a difference).